### PR TITLE
ACSL++ manual

### PR DESCRIPTION
**Pull request by vprevosto**

### CODE REVIEW

**Comment by vprevosto** [comment]
@davidcok I've rebased the branch and merged `main.tex` and `main-cpp.tex`, with a new mechanism to generate the various versions of the document that is less awkward than the `sed`-based approach that we used (see c378632). In addition, there is now a proper draft mode, in which the `\TODO` and `\review` comments will appear. Default is final mode to generate a proper manual (see 07bd0f5). Finally, I've fixed a few issues here and there, as mainly discovered by `make check`. Does it look OK for you?

File: binders.tex [hunk 1]
```diff
@@ -1,13 +1,15 @@
 \begin{syntax}
   binders ::= binder (, binder)* ;
   \
-  binder ::= type-expr variable-ident;
+  binder ::= type-name variable-ident;
```
Comment (vprevosto) [inline binders.tex]: @davidcok this looks suspiciously like #62. Did somehow that branch get merged in this one?

**Comment by davidcok** [comment]
Hi Virgile. The master branch builds correctly for me — thanks for the overall organizational improvements. I will review the text of the documents for typos and formatting problems (and general improvements or corrections) later, but for now this is a good step forward.

- David

**Comment by vprevosto** [comment]
> Hi Virgile. The master branch builds correctly for me

May be I wasn't clear enough, sorry. I haven't merged acslpp2 yet, so that if you want to check if everything builds fine, you have to be on this branch, not master.

**Comment by davidcok** [comment]
Hi - OK - not sure what I was seeing. But I looked at and pushed a few minor changes to the HEAD of acslpp2:
1) Adjusted format of cover and title page to accommodate line overflow in acslpp-implementation
2) The foreword pages (foreword.tex and cpp-foreword.tex) have an incorrect header. Apparently a \chapter* does not have empty header like a \chapter. In fact using \thispagestyle{plain} does not work either - so this slight problem remains.
3) I was seeing confusion in git between VERSION.tex and version.tex . I think I cleaned it up in favor of version.tex, but check that git is correct when you pull the change.

As I already said, I’ll read the bulk of the document for improvements later.

- David

**Comment by vprevosto** [comment]
3) I was seeing confusion in git between VERSION.tex and version.tex . I think I cleaned it up in favor of version.tex, but check that git is correct when you pull the change.

Oops, I forgot about the uppercase version, and I guess that you have a case-insensitive file system that got a little bit confused at this point. Sorry about that. The current `version.tex` is fine anyway.

**Comment by vprevosto** [comment]
@jensgerlach What's your view on the current status of this branch?

**Comment by jensgerlach** [comment]
The examples in 0.11.2, 0.11.3, 0.11.4 are not legal C++ and rejected by a c++ compiler
They look like Java code.
Here are the compiler diagnostics for the Example in 0.11.2.
```
g++    -c -o A.o A.cpp
A.cpp:3:11: error: brackets are not allowed here; to declare an array, place the brackets after the
      name
int[] data;
   ~~     ^
          []
A.cpp:3:7: error: flexible array member 'data' with type 'int []' is not at the end of class
int[] data;
      ^
A.cpp:4:5: note: next field declaration is here
int data_length;
    ^
A.cpp:11:2: error: expected ';' after class
}
 ^
 ;
```

**Comment by vprevosto** [comment]
> The examples in 0.11.2, 0.11.3, 0.11.4 are not legal C++ and rejected by a c++ compiler
> They look like Java code.

Good point. I'll amend the sources to have the examples in independent C++ files that can be checked for compilation errors in the same way we currently handle C examples

**Comment by jensgerlach** [comment]
On page 23 

> programming language type **boolean**

is mentioned. Does this refer to the *programming language*  C++? In this case the name of the type is **bool**.

**Comment by jensgerlach** [comment]
Casts in ACSL and C are usually written as `(type)object`. In C++, the standard prefers imo writing `type(object)`. Should this be reflected to ACSL++?

**Comment by jensgerlach** [comment]
In 2.3.3 (page 36), ACSL++ is mentioned in bold. Should this be set in green?

**Comment by jensgerlach** [comment]
In example 2.24 (page 42) there are no colons after the name of the behaviors.

**Comment by jensgerlach** [comment]
In section 2.4.3 (page 49, line 18 of the second code snippet) there is a `\data.begin`.
I think the slash is wrong there.

I wonder, moreover, if the (corrected) expression 
```
*(data.begin()+j))
```
makes sense in ACSL++, because operator `*` is here an overloaded version for `list<int>::const_iterator`

**Comment by jensgerlach** [comment]
I have noticed that the term _receiver_ is used several times in the context of class methods.
It's not to difficult to understand what is meant but I think this usage is not very common in the C++ community.

**Comment by vprevosto** [comment]
> Good point. I'll amend the sources to have the examples in independent C++ files that can be checked for compilation errors in the same way we currently handle C examples

Done.

**Comment by jensgerlach** [comment]
I have just noticed that I overlooked the description of `\data` in 2.4.3.
Still, I think it's 'definition' and the long example is not so clear.
Maybe one could start with the official defintion of range-based `for` which states that

```
for ( for-range-declaration : for-range-initializer ) statement 
```
is equivalent to
```
{
  auto &&__range = for-range-initializer ; auto __begin = begin-expr ;
  auto __end = end-expr ;
  for ( ; __begin != __end; ++__begin ) {
    for-range-declaration = *__begin;
       statement
  }
}
```

**Comment by jensgerlach** [comment]
Section 2.7.1. An example of a virtual predicate or logic function would be helpful.

**Comment by jensgerlach** [comment]
In the example of 2.7.6 could be a good place to use `\this` instead of `this`.
I would also change the signature of 
```
MyType operator+(MyType const& other);
```
to
```
MyType operator+(const MyType& other) const;
```

**Comment by jensgerlach** [comment]
In 2.7.8 I would change the signatures to
```
 C::C(const C&) = delete;
 C::C(const C&) = default;
```

**Comment by jensgerlach** [comment]
Section 2.7.7. talks about *move constructors* but I could not find any discussion of rvalue references `X&&`.

**Comment by jensgerlach** [comment]
## Section 2.10.4

It would be clearer to me if 
- `hash_code` is replaced by `std::type_info::hash_code`
- `type_index` is replaced by `std::type_index`

**Comment by jensgerlach** [comment]
### 2.11.2
The `Array::getValue` method should be declared as `const`

**Comment by jensgerlach** [comment]
### 2.11.3

The discussion of `constexpr` could be extended (it is only mentioned as a afterthought) in cluding an example.

**Comment by jensgerlach** [comment]
Page 82. There is a typo _Inn particular_

**Comment by jensgerlach** [comment]
### 2.9 Templates

What about templated predicates and logic functions?

**Comment by jensgerlach** [comment]
Section 2.9 speaks of _parameterized template classes_. Maybe _parameterized classes_ suffices.

**Comment by jensgerlach** [comment]
Section 2.21.2. on class invariants is a bit vague.
I thought one would mention here type invariants and their relationship to constructors, destructors and the other member function. I remember that we had quite some discussions about it.

**Comment by jensgerlach** [comment]
Page 113 mentions _Jens_ !-)

**Comment by vprevosto** [comment]
> On page 23
> 
> > programming language type **boolean**
> 
> is mentioned. Does this refer to the _programming language_ C++? In this case the name of the type is **bool**.

In fact, this remark contradicts a later discussion that says that you have an implicit conversion  from bool to boolean but not the other way around, where you must be explicit. I'm just deleting this point.

**Comment by vprevosto** [comment]
> Casts in ACSL and C are usually written as `(type)object`. In C++, the standard prefers imo writing `type(object)`. Should this be reflected to ACSL++?

I'm adding a reference in the bnf for terms to the section where C++/ACSL++ casts are described in more details

**Comment by vprevosto** [comment]
> In 2.3.3 (page 36), ACSL++ is mentioned in bold. Should this be set in green?

No, this is common to ACSL and ACSL++ and speaks about the semantics of the assigns clause.

**Comment by vprevosto** [comment]
> I have noticed that the term _receiver_ is used several times in the context of class methods.
> It's not to difficult to understand what is meant but I think this usage is not very common in the C++ community.

Indeed, the word receiver does not appear at all in the C++11 standard. I'll change the wording

**Comment by davidcok** [comment]
Sorry - though I did lots of C++ prior to Java’s invention, Java has been my primary language for a long time now.

Virgile — is there in place a make target to do the compile checking on these examples?

- David

**Comment by davidcok** [comment]
Oops I see there is a Makefile addition for this purpose

**Comment by davidcok** [comment]
Actually, bool is the C++ type and boolean is the ACSL logic type (like int and integer).

- David

**Comment by davidcok** [comment]
Text revised to note that the comment refers to both ACSL and ACSL++

**Comment by davidcok** [comment]
Done.

**Comment by davidcok** [comment]
Done,

**Comment by davidcok** [comment]
Fixed.

**Comment by davidcok** [comment]
Actually - which would be preferable?  
a) A distinction between bool and boolean, with casts required in some cases
or
b) A distinction, but no casts required
or
c) Equivalence - no casts and interchangeability.

I prefer (a) or (b) as it helps the reader remember to distinguish logic expressions from programming language expressions.

- David

**Comment by vprevosto** [comment]
> Actually - which would be preferable? a) A distinction between bool and boolean, with casts required in some cases or b) A distinction, but no casts required or c) Equivalence - no casts and interchangeability. I prefer (a) or (b) as it helps the reader remember to distinguish logic expressions from programming language expressions. - David

I prefer a) as well (implicit conversion from C++ to ACSL++, cast required the other way around)

**Comment by davidcok** [comment]
Added some text and grammar material.

**Comment by davidcok** [comment]
Changed as suggested.

**Comment by davidcok** [comment]
Add an item in the TODO list, as this point will take some thinking and design.

**Comment by davidcok** [comment]
Jens - are there particular issues you suggest be addressed? You likely are more familiar with uses of constexpr than I am.

- David

**Comment by davidcok** [comment]
I thought these were not necessary because ACSL already includes the ability to create logic type, functions and predicates with type parameters. There seems no need to replicate this in the C++ style.

Do you concur or not?

- David

**Comment by davidcok** [comment]
And I added a grammar figure.

**Comment by davidcok** [comment]
In fact the problem is using the C++ variable data and the logic variable \data in the same example, as it leads to this confusion.

\data (and \count) are introduced as logic quantities to be able to state things that the program typically cannot.

I revised the example.

- David

**Comment by davidcok** [comment]
I’ll  leave that point to be dealt with as the TODO items are resolved.

- David

**Pull request merged**